src/query/block/scalar.go

```diff
@@ -0,0 +1,172 @@
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package block
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/m3db/m3/src/query/models"
+)
+
+// ScalarBlock is a block containing a single value
+// over a certain bound
+type ScalarBlock struct {
+	val  float64
+	meta Metadata
+}
+
+// assert that ScalarBlock implements Block
+var _ Block = (*ScalarBlock)(nil)
+
+// NewScalarBlockResult creates a Result consisting
+// of scalar blocks over the bound
+func NewScalarBlockResult(val float64, bounds Bounds) *Result {
+	blocks := []Block{
+		NewScalarBlock(val, bounds),
+	}
+
+	return &Result{
+		Blocks: blocks,
+	}
+}
+
+// NewScalarBlock creates a scalar block containing val over the bounds
+func NewScalarBlock(val float64, bounds Bounds) Block {
+	return &ScalarBlock{
+		val: val,
+		meta: Metadata{
+			Bounds: bounds,
+			Tags:   make(models.Tags),
+		},
+	}
+}
+
+// StepIter returns a StepIterator
+func (b *ScalarBlock) StepIter() (StepIter, error) {
+	bounds := b.meta.Bounds
+	steps := bounds.Steps()
+	return &scalarStepIter{
+		meta: b.meta,
+		step: scalarStep{
+			vals: []float64{b.val},
+		},
+		numVals: steps,
+		idx:     -1,
+	}, nil
+}
+
+// SeriesIter returns a SeriesIterator
+func (b *ScalarBlock) SeriesIter() (SeriesIter, error) {
+	bounds := b.meta.Bounds
+	steps := bounds.Steps()
+	vals := make([]float64, steps)
+	for i := range vals {
+		vals[i] = b.val
+	}
+	return &scalarSeriesIter{
+		meta: b.meta,
+		vals: vals,
+		idx:  -1,
+	}, nil
+}
+
+// Close closes the scalar block
+func (b *ScalarBlock) Close() error { return nil }
+
+// Value returns the value for the scalar block
+func (b *ScalarBlock) Value() float64 { return b.val }
+
+type scalarStepIter struct {
+	meta    Metadata
+	step    scalarStep
+	numVals int
+	idx     int
+}
+
+// assert that scalarStepIter implements StepIter
+var _ StepIter = (*scalarStepIter)(nil)
+
+var seriesMeta = []SeriesMeta{
+	SeriesMeta{
+		Tags: models.Tags{},
+	},
+}
+
+func (it *scalarStepIter) Close()                   { /* No-op*/ }
+func (it *scalarStepIter) StepCount() int           { return it.numVals }
+func (it *scalarStepIter) SeriesMeta() []SeriesMeta { return seriesMeta }
+func (it *scalarStepIter) Meta() Metadata           { return it.meta }
+
+func (it *scalarStepIter) Next() bool {
+	it.idx++
+	bounds := it.meta.Bounds
+	it.step.time = bounds.Start.Add(time.Duration(it.idx) * bounds.StepSize)
+	return it.idx < it.numVals
+}
+
+func (it *scalarStepIter) Current() (Step, error) {
+	if it.idx >= it.numVals || it.idx < 0 {
```

Comment:
we haven't traditionally been checking this in Current() usually rely on people to call Next() before Current

Reply:
I agree in general, but since this doesn't do any real "bounds" checking and just returns a value, the alternative is that this `panics` in this check, as it really shouldn't be returning anything if Next() is false.

I prefer to error than to panic here, but if you have particular preference towards panicing I can change to that

Reply:
other places have panics instead of errors since its a way to avoid more codepaths :( I guess we can do either here since its returning error but a lot of times Current() only returns data and panic is the only option.

Reply:
Yeah, if it weren't already able to return an error I would've paniced, but figured might as well make use of the capability here so we don't kill the node :)